### PR TITLE
Ports /vg/s Smoothwall code

### PR DESCRIPTION
**Pull request by RemieRichards**
- Smoothwalling is now faster
- Smoothwalling is now OOP and extensible
- Added the ability to stop a wall from smoothing during laggy times, unused at the moment
- Angry comments from N3X15 from beyond the grave

Fixes #lag

### CODE REVIEW

**Comment by WJohn** [comment]
Speeding the game up is what you're in charge of now, apparently.

I'm okay with that.

**Comment by RemieRichards** [comment]
@WJohn that and random things relating to pixels:
* get_pixel_distance()
* DrawPixel()
* DrawPixelOn() 
* get_turf_pixel()

But yeah:
* Explosion performance
* Image caches

**Comment by Cheridan** [comment]
>Added the ability to stop a wall from smoothing during laggy times

Phew, this should help a lot with our super laggy smoothwall code. I'm glad someone understands smoothwalls.

**Comment by Razharas** [comment]
Damn cant understand a single line

**Comment by RemieRichards** [comment]
@Razharas you'll get there, I believe in you

**Comment by xxalpha** [comment]
Do the walls ever smooth after the singularity is gone?

**Comment by RemieRichards** [comment]
@xxalpha yes, because I didn't include that use case, since we don't really lag with smoothwalls.
if I added it back, no, they would never smooth again until something nearby them also smooth (new wall built, wall destroyed etc.)

**Comment by Fox-McCloud** [comment]
Hopefully this is updated/fixed to the point where walls actually smooth.

We had this, for a while, and it resulted in walls not smooth for replaced walls, but walls smoothing for walls adjacent to replaced walls.

Either case, this is probably fixed, but just a heads up.

**Comment by RemieRichards** [comment]
@Fox-McCloud we don't have those issues, mainly because N3X15 did eventually add 2 fix PRs.

**Comment by Fox-McCloud** [comment]
@RemieRichards 

Great to hear; always hard to keep up with all the little details of so many codebases. Thanks!

File: code/game/smoothwall.dm [hunk 1]
```diff
@@ -1,112 +1,87 @@
...
-				src.icon = newicon; //we'll set the floor's icon to the floor next to it and overlay the wall segment. shuttle floor sprites have priority
-				src.icon_state = newiconstate; //
-				src.overlays += icon('icons/turf/shuttle.dmi',"swall_f[junction]")
-			else //otherwise, the space one
-				src.icon_state = "swall_s[junction]"*/
-
-	return
+// OKAY I DON'T KNOW WHO THE FUCK ORIGINALLY CODED THIS BUT THEY ARE OFFICIALLY FIRED FOR BEING DRUNK AND STUPID
+// FUCK YOU MYSTERY CODERS
+// FOR THIS SHIT I'M GOING TO MAKE ALL MY COMMENTS IN CAPS
```
Comment (Miauw62) [inline code/game/smoothwall.dm]: you should probably add a note here that these comments were made by nexis :V

**Comment by Supermichael777** [comment]
Clown false wall needs to make a bikehorn sound when it opens.

File: code/game/smoothwall.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/smoothwall.dm]: well I did in the PR~ Haha
It's kinda nice for people who don't know the story to be like "Wow, this guy's angry!"

**Comment by optimumtact** [comment]
merge?

**Comment by RemieRichards** [comment]
@tgstation/commit-access mergy mergy pls pls

**Comment by optimumtact** [comment]
BYLAT MERGE @Cheridan

File: code/game/smoothwall.dm [hunk 2]
```diff
@@ -1,112 +1,85 @@
...
+
+/atom
+	var/list/canSmoothWith=list() // TYPE PATHS I CAN SMOOTH WITH~~~~~
```
Comment (Aranclanos) [inline code/game/smoothwall.dm]: i hate this

**Pull request merged**